File: Leetcode2023/March2023/C++/ScrambleString.cpp
// https://leetcode.com/problems/scramble-string/
class Solution {
private:
    bool helper(string s1, string s2, unordered_map<string, bool>& dp) {
        // base case
        if (s1 == s2) {
            return true;
        }
        if (s1.length() <= 1 || s2.length() <= 1) {
            return false;
        }
        
        string key = s1 + "-" + s2;
        if (dp.count(key)) {
            return dp[key];
        }
        
        int n = s1.length();
        bool flag = false;
        
        for (int i = 1; i <= n - 1; i++) {
            bool swap = helper(s1.substr(0, i), s2.substr(n - i), dp) && helper(s1.substr(i), s2.substr(0, n - i), dp);
            if (swap) {
                flag = true;
                break;
            }
            
            bool notSwap = helper(s1.substr(0, i), s2.substr(0, i), dp) && helper(s1.substr(i), s2.substr(i), dp);
            if (notSwap) {
                flag = true;
                break;
            }
        }
                
        return dp[key] = flag;
    }
    
public:
    bool isScramble(string s1, string s2) {
        unordered_map<string, bool> dp;
        return helper(s1, s2, dp);
    }
};

File: March2023/Java/ScrambleString.java
class Solution {
    public boolean isScramble(String s1, String s2) {
        Map<String, Boolean> dp = new HashMap<>();
        return helper(s1, s2, dp);
    }
    
    private boolean helper(String s1, String s2, Map<String, Boolean> dp) {
        // base case
        if (s1.equals(s2)) {
            return true;
        }
        if (s1.length() <= 1 || s2.length() <= 1) {
            return false;
        }
        
        String key = s1 + "-" + s2;
        if (dp.containsKey(key)) {
            return dp.get(key);
        }
        
        int n = s1.length();
        boolean flag = false;
        
        for (int i = 1; i <= n - 1; i++) {
            boolean swap = helper(s1.substring(0, i), s2.substring(n - i), dp) && helper(s1.substring(i), s2.substring(0, n - i), dp);
            if (swap) {
                flag = true;
                break;
            }
            
            boolean notSwap = helper(s1.substring(0, i), s2.substring(0, i), dp) && helper(s1.substring(i), s2.substring(i), dp);
            if (notSwap) {
                flag = true;
                break;
            }
        }
        
        dp.put(key, flag);
        return flag;
    }
}
